File: Swift 풀이/DP/[백준 1912] 연속합.swift
//
//  [백준 1912] 연속합.swift
//  Algorithm
//
//  Created by 김민 on 2/23/24.
//
// 백준 1912 연속합 https://www.acmicpc.net/problem/1912

/*
[📌 전략][✅ 풀이]
dp[i]: i번째 인덱스의 수를 마지막으로 하는 연속하는 수들의 최대 합
입력되는 수열을 nums라 하면, dp[i]는
1. nums[i]를 마지막으로 포함하여 연속적으로 선택을 했을 때의 최대 합 = dp[i-1] + nums[i]
2. 연속적으로 선택하지 않고 nums[i]만 선택했을 때 = nums[i]
총 2가지 경우가 올 수 있고 이중 큰 값으로 dp[i]에 저장
최종 답은 dp 값들 중 최댓값이 된다.
*/

let n = Int(readLine()!)!
let nums = readLine()!.split(separator: " ").map { Int($0)! }
var dp = Array(repeating: 0, count: n)

for i in 0..<n {
    if i == 0 {
        dp[0] = nums[0]
        continue
    }

    dp[i] = max(dp[i-1] + nums[i], nums[i])
}

print(dp.max()!)

File: Python/DP/[백준 1912] 연속합.py
'''
[백준 1912] 연속합
https://www.acmicpc.net/problem/1912

[📌 전략]
- dp로 풀이

[✅ 풀이]
- nums 입력받음
- nums를 순회하며 이전 값을 현재 값에 더할 것
    - 단, 더한 값이 현재 값보다 클 때만 더해서 연속적인 최대 합을 구함

[📝 기록]
LV: 실버3
풀이 시간: 27분 
'''

import sys

input = sys.stdin.readline

n = int(input())
nums = list(map(int, input().split()))

for i in range(1, n):
    nums[i] = max(nums[i-1] + nums[i], nums[i])

print(max(nums))